README.md
# Cimpress Fulfillment Location client

### Usage
In order to use the client, install the package by running:
```
npm install cimpress-fulfillment-location --save
```

Once the package is available, you can retrieve a specific fulfillment location as such:
```
const fulfillmentLocationId = 'abcd1234';
let client = new FulfillmentLocationClient({
    log: defaultLogger,
    cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 },
    timeout: 2000
});

client.getLocation(fulfillmentLocationId, {
    accessToken: "Your access token",
    skipCache: false, // Skips the cache so the results will be fresh, if not set is false by default
})
    .then(fulfillmentLocation => {
        // do stuff
    });
```

Otherwise, you can retrieve a list of all the fulfillment locations you have access to like this:
```
let client = new FulfillmentLocationClient({
    log: defaultLogger
});

client.getLocations({
    accessToken: "Your access token",
    fulfillerId: "Your fulfiller Id", // Optional
    showArchived: false, // By default is false, shows archived fulfillment locations
    skipCache: false // Skips the cache so the results will be fresh, if not set is false by default
})
    .then(fulfillmentLocations => {
        // do stuff
    });

```

### Prerequisites
* [Node.js](https://nodejs.org/en/)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.8] - 2022-01-07
### Fix
- Fix incomplete package deployment

## [1.0.7] - 2022-01-07
### Fix
- Remove dependency on non-standard Error.captureStackTrace() method

## [1.0.6] - 2021-03-16
### Fix
- Applying security vulnerabilities patches

## [1.0.5] - 2020-08-28
### Changed
- Move to Node.js major version 12

### Fixed
- On error from Fulfillment Location, fix the returned name of the error

## [1.0.4] - 2020-01-23
### Add
- Add the posibility of filtering fulfillment locations by fulfiller

## [1.0.3] - 2019-09-10
### Fix
- Applying security vulnerabilities patches

## [1.0.2] - 2019-06-23
### Fix
- Updating axios and bunch of dev dependencies to reduce vulnerabilities

## [1.0.1] - 2018-11-29
### Fix
- Solve critical vulnerability caused by growl, a dependency of mocha.

## [1.0] - 2018-11-29
### Changed
- Aditional properties of getLocations and getLocation moved to a single object paramenter `options`. 

### Added
- Added new skipCache option that disables all the caches.

## [0.2] - 2018-10-03
### Added
- Enable showArchived flag for listing fulfillment locations.


## [0.1] - 2017-10-11
### Added
- Implemented client for Fulfillment Location service, with caching.






tests/indexTests.js
'use strict';

const expect = require('chai').expect;

describe('Index', () => {

    it('Exports correctly', () => {

        let exported = require('../src/index');
        expect(exported).to.deep.equal({
            FulfillmentLocationClient: require('../src/fulfillmentLocation')
        });

    });

});

tests/fulfillmentLocationTests.js
'use strict';

const expect = require('chai').expect;
const nock = require('nock');
const sinon = require('sinon');

const FulfillmentLocationClient = require('../src/fulfillmentLocation');

// ----- Helpers -----

const defaultLogger = () => ({
    info: sinon.stub(),
    error: sinon.stub()
});

const sampleLocations = () => (
    [
        {
            timeZone: "Europe/London",
            internalFulfillerId: 70,
            internalFulfillmentLocationId: 110,
            fulfillerId: "a2wgr294u",
            fulfillmentLocationId: "b7uqagcx0nw"
        },
        {
            timeZone: "Europe/Paris",
            internalFulfillerId: 70,
            internalFulfillmentLocationId: 189,
            fulfillerId: "a2wgr294u",
            fulfillmentLocationId: "bqcjg7qbvep"
        }
    ]
);

// ----- END -----

describe('getLocations :: without cache ::', function () {

    it('FL returns 200 :: returns list of fulfillment locations correctly', function () {
        nock.cleanAll();

        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations")
            .times(1)
            .reply(200, sampleLocations());

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocations('Bearer X')
            .then(data => {
                expect(data).to.deep.equal(sampleLocations());
            });
    });

    it('FL returns 500 :: returns an error', function () {
        nock.cleanAll();

        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations")
            .times(1)
            .reply(500, 'Nah, not working right now');

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocations('Bearer X')
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.response.status).to.equal(500);
                expect(error.response.data).to.equal('Nah, not working right now');
            });
    });
});

describe('getLocations :: with cache ::', function () {
    
    it('FL returns 200 :: returns list of fulfillment locations correctly', function () {
        nock.cleanAll();

        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations")
            .times(1)
            .reply(200, sampleLocations());

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocations('Bearer X')
            .then(data => {
                expect(data).to.deep.equal(sampleLocations());
            });
    });

    it('locations are in the cache and uses the value in the cache', function () {
        nock.cleanAll();

        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations")
            .times(1)
            .reply(200, sampleLocations());

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocations('Bearer X')
            .then(_data => {
                nock('https://fulfillmentlocation.trdlnk.cimpress.io')
                    .get("/v1/fulfillmentlocations")
                    .times(1)
                    .reply(500, 'Unable to load locations');

                client.getLocations('Bearer X')
                    .then(data => {
                        expect(data).to.equal(sampleLocations());
                    })
                    .catch(error => {
                        expect(error).to.not.exist;
                    });
            });
    });

    it('FL returns 500 :: returns an error', function () {
        nock.cleanAll();

        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations")
            .times(1)
            .reply(500, 'Nah, not working right now');

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocations('Bearer X')
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.response.status).to.equal(500);
                expect(error.response.data).to.equal('Nah, not working right now');
            });
    });
});

describe('getLocation :: without cache ::', function () {

    afterEach(() => {
        nock.cleanAll();
    });

    it('FL service returns 200 and the location requested (using alphanum id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/bqcjg7qbvep")
            .reply(200, sampleLocations()[1]);

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation('Bearer X', "bqcjg7qbvep")
            .then(data => {
                expect(data).to.deep.equal(sampleLocations()[1]);
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('FL service returns 404 for invalid alphanum id', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/a7uqagcx0nz")
            .reply(404, "Location 'a7uqagcx0nz' not found");

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation('Bearer X', "a7uqagcx0nz")
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.response.status).to.equal(404);
                expect(error.response.data).to.equal(`Location 'a7uqagcx0nz' not found`);
            });
    });

    it('FL service returns 200 and the location requested (using internal id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/189")
            .reply(200, sampleLocations()[1]);

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation('Bearer X', "189")
            .then(data => {
                expect(data).to.deep.equal(sampleLocations()[1]);
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('FL service returns 404 for invalid internal id', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/180")
            .reply(404, "Location '180' not found");

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation('Bearer X', "180")
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.response.status).to.equal(404);
                expect(error.response.data).to.equal(`Location '180' not found`);
            });
    });
});

describe('getLocation :: with cache ::', function () {
    
    afterEach(() => {
        nock.cleanAll();
    });

    it('FL service returns 200 and the location requested (using alphanum id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/bqcjg7qbvep")
            .reply(200, sampleLocations()[1]);

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation('Bearer X', "bqcjg7qbvep")
            .then(data => {
                expect(data).to.deep.equal(sampleLocations()[1]);
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('location is in cache and uses the value in the cache', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/bqcjg7qbvep")
            .reply(200, sampleLocations()[1]);

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client
            .getLocation('Bearer X', "bqcjg7qbvep")
            .then(_data => {
                // FL service is now returning a 404; client should use data in cache
                nock('https://fulfillmentlocation.trdlnk.cimpress.io')
                    .get("/v1/fulfillmentlocations/bqcjg7qbvep")
                    .reply(500, "Unable to load information for 'bqcjg7qbvep'");

                client
                    .getLocation('Bearer X', "bqcjg7qbvep")
                    .then(data => {
                        expect(data).to.deep.equal(sampleLocations()[1]);
                    })
                    .catch(error => {
                        expect(error).to.not.exist;
                    });
            });
    });

    it('FL service returns 404 for invalid alphanum id', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/a7uqagcx0nz")
            .reply(404, "Location 'a7uqagcx0nz' not found");

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation('Bearer X', "a7uqagcx0nz")
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.response.status).to.equal(404);
                expect(error.response.data).to.equal(`Location 'a7uqagcx0nz' not found`);
            });
    });

    it('FL service returns 200 and the location requested (using internal id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/189")
            .reply(200, sampleLocations()[1]);

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation('Bearer X', "189")
            .then(data => {
                expect(data).to.deep.equal(sampleLocations()[1]);
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('FL service returns 404 for invalid internal id', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/180")
            .reply(404, "Location '180' not found");

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation('Bearer X', "180")
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.response.status).to.equal(404);
                expect(error.response.data).to.equal(`Location '180' not found`);
            });
    });
});

tests/getFulfillmentLocationTests.js
'use strict';

const expect = require('chai').expect;
const nock = require('nock');
const sinon = require('sinon');

const FulfillmentLocationClient = require('../src/fulfillmentLocation');

// ----- Helpers -----

const defaultLogger = () => ({
    info: sinon.stub(),
    error: sinon.stub()
});

const sampleLocation = () => (
    {
        timeZone: "Europe/London",
        internalFulfillerId: 70,
        internalFulfillmentLocationId: 110,
        fulfillerId: "a2wgr294u",
        fulfillmentLocationId: "b7uqagcx0nw"
    }
);

// ----- END -----

describe('getLocation :: without cache ::', function () {

    afterEach(() => {
        nock.cleanAll();
    });

    it('FL service returns 200 and the location requested (using alphanum id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/bqcjg7qbvep")
            .reply(200, sampleLocation());

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation("bqcjg7qbvep", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.deep.equal(sampleLocation());
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('FL service :: skipCache set :: returns 200 and the fresh location requested (using alphanum id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io', {reqheaders: {
            'Cache-Control': 'no-cache',
            'X-Cache-Id': (value) => value!= null
        }})
            .get("/v1/fulfillmentlocations/bqcjg7qbvep")
            .reply(200, sampleLocation());

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation("bqcjg7qbvep", {accessToken: 'Bearer X', skipCache: true})
            .then(data => {
                expect(data).to.deep.equal(sampleLocation());
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('FL service returns 404 for invalid alphanum id', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/a7uqagcx0nz")
            .reply(404, "Location 'a7uqagcx0nz' not found");

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation("a7uqagcx0nz", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.status).to.equal(404);
                expect(error.additionalData).to.equal(`Location 'a7uqagcx0nz' not found`);
            });
    });

    it('FL service returns 200 and the location requested (using internal id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/189")
            .reply(200, sampleLocation());

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation("189", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.deep.equal(sampleLocation());
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('FL service returns 404 for invalid internal id', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/180")
            .reply(404, "Location '180' not found");

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation("180", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.status).to.equal(404);
                expect(error.additionalData).to.equal(`Location '180' not found`);
            });
    });

    it('FL service returns 401 for unauthorized user', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/180")
            .reply(401);

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation("180", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.status).to.equal(401);
            });
    });

    it('FL service returns 403 for forbidden', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/180")
            .reply(403);

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation("180", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.status).to.equal(403);
            });
    });

    it('Authorization parameter is not passed :: returns an error', function () {
        nock.cleanAll();

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation("180")
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.message).to.equal('Missing Authorization parameter');
            });
    });

    it('Authorization parameter is of an invalid format :: returns an error', function () {
        nock.cleanAll();

        let client = new FulfillmentLocationClient({log: defaultLogger()});

        return client.getLocation("180", {accessToken: 'invalid-format'})
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.message).to.equal('Invalid format for Authorization parameter: "invalid-format". Authorization parameter should be in the following format: "Bearer [token]"');
            });
    });
});

describe('getLocation :: with cache ::', function () {
    
    afterEach(() => {
        nock.cleanAll();
    });

    it('FL service returns 200 and the location requested (using alphanum id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/bqcjg7qbvep")
            .reply(200, sampleLocation());

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation("bqcjg7qbvep", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.deep.equal(sampleLocation());
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('FL service :: skipCache set :: rreturns 200 and the fresh location requested (using alphanum id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io', {reqheaders: {
            'Cache-Control': 'no-cache',
            'X-Cache-Id': (value) => value!= null
        }})
            .get("/v1/fulfillmentlocations/bqcjg7qbvep")
            .reply(200, sampleLocation());


        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation("bqcjg7qbvep", {accessToken: 'Bearer X', skipCache: true})
            .then(data => {
                expect(data).to.deep.equal(sampleLocation());
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('FL service returns 404 for invalid alphanum id', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/a7uqagcx0nz")
            .reply(404, "Location 'a7uqagcx0nz' not found");

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation("a7uqagcx0nz", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.status).to.equal(404);
                expect(error.additionalData).to.equal(`Location 'a7uqagcx0nz' not found`);
            });
    });

    it('FL service returns 200 and the location requested (using internal id)', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/189")
            .reply(200, sampleLocation());

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation("189", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.deep.equal(sampleLocation());
            })
            .catch(error => {
                expect(error).to.not.exist;
            });
    });

    it('FL service returns 404 for invalid internal id', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/180")
            .reply(404, "Location '180' not found");

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation("180", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.status).to.equal(404);
                expect(error.additionalData).to.equal(`Location '180' not found`);
            });
    });

    it('FL returns 500 :: location is in cache and uses the value in the cache', function () {
        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/bqcjg7qbvep")
            .reply(200, sampleLocation());

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client
            .getLocation("bqcjg7qbvep", {accessToken: 'Bearer X'})
            .then(_data => {
                // FL service is now returning a 404; client should use data in cache
                nock('https://fulfillmentlocation.trdlnk.cimpress.io')
                    .get("/v1/fulfillmentlocations/bqcjg7qbvep")
                    .reply(500, "Unable to load information for 'bqcjg7qbvep'");

                client
                    .getLocation("bqcjg7qbvep", {accessToken: 'Bearer X'})
                    .then(data => {
                        expect(data).to.deep.equal(sampleLocation());
                    })
                    .catch(error => {
                        expect(error).to.not.exist;
                    });
            });
    });

    it('FL returns 500 :: returns an error', function () {
        nock.cleanAll();

        nock('https://fulfillmentlocation.trdlnk.cimpress.io')
            .get("/v1/fulfillmentlocations/bqcjg7qbvep")
            .reply(500, "Unable to load information for 'bqcjg7qbvep'");

        let client = new FulfillmentLocationClient({
            log: defaultLogger(),
            cacheConfig: { stdTTL: 4 * 60 * 60, checkperiod: 5 * 60 }
        });

        return client.getLocation("bqcjg7qbvep", {accessToken: 'Bearer X'})
            .then(data => {
                expect(data).to.not.exist;
            })
            .catch(error => {
                expect(error.response.status).to.equal(500);
                expect(error.response.data).to.equal("Unable to load information for 'bqcjg7qbvep'");
            });
    });
});

src/fulfillmentLocation.js
'use strict';

const DEFAULT_URL = "https://fulfillmentlocation.trdlnk.cimpress.io";

const nodeCache = require('node-cache');
let flCache;
let axios = require('axios');

// --- Predefined errors ---

function UnauthorizedError(message, extra) {
    let error = new Error(message || 'Unauthorized');
    error.name = 'UnauthorizedError';
    error.status = 401;
    error.additionalData = extra;
    return error;
}

function ForbiddenError(message, extra) {
    let error = new Error(message || 'Forbidden');
    error.name = 'ForbiddenError';
    error.status = 403;
    error.additionalData = extra;
    return error;
}

function NotFoundError(message, extra) {
    let error = new Error(message || 'Not found');
    error.name = 'NotFoundError';
    error.status = 404;
    error.additionalData = extra;
    return error;
}

// --- END ---

const handleAuthorization = (authorization, reject) => {
    if ( !authorization ) {
        return reject(new Error('Missing Authorization parameter'));
    }

    if ( authorization.indexOf('Bearer ') !== 0 ) {
        return reject(new Error(`Invalid format for Authorization parameter: "${authorization}". Authorization parameter should be in the following`
            + ` format: "Bearer [token]"`));
    }
};

const handleError = (err, reject) => {
    if ( err.status === 401 || (err.response && err.response.status === 401) ) {
        return reject(new UnauthorizedError(err.statusText || err.message, err.data || err.response.data));
    }
    if ( err.status === 403 || (err.response && err.response.status === 403) ) {
        return reject(new ForbiddenError(err.statusText || err.message, err.data || err.response.data));
    }
    if ( err.status === 404 || (err.response && err.response.status === 404) ) {
        return reject(new NotFoundError(err.statusText || err.message, err.data || err.response.data));
    }
    return reject(err);
};

class FulfillmentLocationClient {

    constructor(c) {
        let config = c || {};
        this.url = config.url || DEFAULT_URL;
        this.log = config.log;
        this.useCaching = !!config.cacheConfig;
        if ( config.cacheConfig ) {
            flCache = new nodeCache(config.cacheConfig);
        }
        this.timeout = config.timeout || 2000;
    }

    getLocation(locationId, options) {

        return new Promise((resolve, reject) => {
            const opts = options || {};
            handleAuthorization(opts.accessToken, reject);
            const instance = axios.create({
                baseURL: this.url,
                timeout: this.timeout
            });

            instance.defaults.headers.common['Authorization'] = opts.accessToken;

            if ( !this.useCaching || opts.skipCache) {
                return this._getFulfillmentLocationFromService(instance, locationId, opts)
                    .then(location => {
                        return resolve(location);
                    })
                    .catch(err => {
                        return reject(err);
                    });
            }

            const cacheKey = 'fulfillmentLocation_' + locationId + '_' + opts.accessToken;
            flCache.get(cacheKey, (err, fulfillmentLocation) => {

                if ( err ) {
                    this.log.error(`Failed to read from node-cache (key=${cacheKey})`, err);
                }

                if ( err || (fulfillmentLocation == undefined) ) {

                    return this._getFulfillmentLocationFromService(instance, locationId, opts)
                        .then(location => {
                            flCache.set(cacheKey, location);
                            return resolve(location);
                        })
                        .catch(err => {
                            return reject(err);
                        });

                } else {
                    return resolve(fulfillmentLocation);
                }

            });
        });
    }

    getLocations(options) {

        return new Promise((resolve, reject) => {
            const opts = options || {};
            handleAuthorization(opts.accessToken, reject);
            const instance = axios.create({
                baseURL: this.url,
                timeout: this.timeout
            });
    
            instance.defaults.headers.common['Authorization'] = opts.accessToken;
    
            if ( !this.useCaching || opts.skipCache) {
                return this._getFulfillmentLocationsFromService(instance, opts)
                    .then(locations => {
                        return resolve(locations);
                    })
                    .catch(err => {
                        return reject(err);
                    });
            }

            const cacheKey = 'fulfillmentLocations_' + opts.accessToken;
            flCache.get(cacheKey, (err, fulfillmentLocations) => {

                if ( err ) {
                    this.log.error(`Failed to read from node-cache (key=${cacheKey})!?`, err);
                }

                if ( err || (fulfillmentLocations == undefined) ) {

                    return this._getFulfillmentLocationsFromService(instance, opts)
                        .then(locations => {
                            flCache.set(cacheKey, locations);
                            return resolve(locations);
                        })
                        .catch(err => {
                            return reject(err);
                        });

                } else {
                    return resolve(fulfillmentLocations);
                }

            });
        });
    }

    _getFulfillmentLocationFromService(instance, fulfillmentLocationId, options) {
        return new Promise((resolve, reject) => {
            const endpoint = this.url + "/v1/fulfillmentlocations/" + fulfillmentLocationId;
            const requestConfig = {
                method: 'GET',
                url: endpoint,
                timeout: this.timeout,
                headers: {}
            };
            
            if (options.skipCache){
                requestConfig.headers['Cache-Control'] = 'no-cache';
                requestConfig.headers['X-Cache-Id'] = Math.random();
            }

            if ( this.log && this.log.info ) {
                this.log.info("->" + endpoint, requestConfig);
            }

            instance
                .request(requestConfig)
                .then(res => {
                    
                    if ( this.log && this.log.info ) {
                        this.log.info("<-" + endpoint, res.data);
                    }

                    return resolve(res.data || []);
                })
                .catch(err => {
                    if ( this.log && this.log.error ) {
                        this.log.error("<-" + endpoint, err);
                    }

                    return handleError(err, reject);
                });
        });
    }

    _getFulfillmentLocationsFromService(instance, options) {
        return new Promise((resolve, reject) => {
            const endpoint = this.url + "/v1/fulfillmentlocations";
            const requestConfig = {
                method: 'GET',
                url: endpoint,
                params:{
                    showArchived: options.showArchived || false
                },
                headers: {},
                timeout: this.timeout
            };

            if (options.fulfillerId) {
                requestConfig.params.fulfillerId = options.fulfillerId;
            }

            if (options.skipCache){
                requestConfig.headers['Cache-Control'] = 'no-cache';
                requestConfig.headers['X-Cache-Id'] = Math.random();
            }

            if ( this.log && this.log.info ) {
                this.log.info("->" + endpoint, requestConfig);
            }
            
            instance
                .request(requestConfig)
                .then((res) => {

                    if ( this.log && this.log.info ) {
                        this.log.info("<-" + endpoint, res.data);
                    }

                    return resolve(res.data || []);
                })
                .catch((err) => {

                    if ( this.log && this.log.error ) {
                        this.log.error("<-" + endpoint, err);
                    }

                    return handleError(err, reject);
                });
        });
    }
}

module.exports = FulfillmentLocationClient;
